core/store/src/trie/shard_tries.rs

```diff
@@ -412,8 +391,8 @@ impl ShardTries {
     /// Note that flat storage needs to be handled separately
     pub fn delete_trie_for_shard(&self, shard_uid: ShardUId, store_update: &mut StoreUpdate) {
         // Clear both caches and remove state values from store
-        self.0.caches.write().expect(POISONED_LOCK_ERR).remove(&shard_uid);
-        self.0.view_caches.write().expect(POISONED_LOCK_ERR).remove(&shard_uid);
+        let _cache = self.0.caches.lock().expect(POISONED_LOCK_ERR).remove(&shard_uid);
+        let _view_cache = self.0.view_caches.lock().expect(POISONED_LOCK_ERR).remove(&shard_uid);
```

Comment:
```suggestion
        self.0.caches.lock().expect(POISONED_LOCK_ERR).remove(&shard_uid);
        self.0.view_caches.lock().expect(POISONED_LOCK_ERR).remove(&shard_uid);
```

Reply:
FWIW this is intentional. With how Rust's lifetime extension within a statement works, the `drop_in_place` for the removed value would get executed before the lock guard itself would be released. With this change the order of operations is reversed and the lock is released before we move on to clean up the cache structure.